### ISSUE TITLE
PaymentTransactionHandler.paymentQueue() doesn't return PurchasedItem for auto renewable subscription. Error: Fatal Error - Unexpectedly found nil while unwrapping on Optional value

### ISSUE THREAD

**Issue by vinvantest** [dooboolab/flutter_inapp_purchase#28]
### Version of flutter_inapp_purchase
flutter_inapp_purchase: "^0.6.4"

### Platforms you faced the error (IOS or Android or both?)
iOS

### Expected behavior

On auto renewable subscription purchase the control to return back to Dart code from where buySubscription() is called. Once the subscription is complete (which actually does... it prompts all the payment dialog boxes from apple while subscribing to the auto renewable subscription) the intention is to get the PurchasedItem purchased.transactionReceipt and send it to server for verification.

### Actual behavior (error)

The subscription purchase on apple end seems to go smooth. I get confirmation on the screen "You are all set. Your purchase was successful. [Environment Sandbox]". In Xcode console I can see purchase complete. Log trace below

 Purchase Started !!
Transaction in progress: <SKPaymentTransaction: ########>
2018-09-05 06:02:12.584199+1000 Runner[1456:557945] [App] if we're in the real pre-commit handler we can't actually add any new fences due to CA restriction
 Purchase Successful !! 
.
Transaction purchased or restored: <SKPaymentTransaction: same as ##### above>
warning: could not execute support code to read Objective-C class data in the process. This may reduce the quality of type information available.

**<nothing is printed after this statement>**

THE ERROR --> And in Xcode I get for this code snippet error indicator

func paymentQueue(_ queue: SKPaymentQueue, updatedTransactions transactions: [SKPaymentTransaction]) {
        var payments:Array<Dictionary<String, Any>> = []
        
        for transaction in transactions {
            switch transaction.transactionState {
            case .failed:
                queue.finishTransaction(transaction)
                print("Transaction Failed \(transaction)")
            case .purchased, .restored:
                queue.finishTransaction(transaction)
                print("Transaction purchased or restored: \(transaction)")
                payments.append(paymentToDictionary(transaction))
            case .deferred, .purchasing:
                print("Transaction in progress: \(transaction)")
            }
        }
        
        if(payments.count > 0) {
            flutterResult!(payments)      **<-- HERE I GET THE ERROR**
            flutterResult = nil
        }
    } 

The error in xcode is Error: Fatal Error - Unexpectedly found nil while unwrapping on Optional value on the line indicated above.

I was able to view the payments object and it has 1 item with 4 key , values. Xcode console printed them
"1000000########"
"#####_monthly_basic"
"#####_monthly_basic"
Printing description of ((Swift.Int)1536091######):
15360913#####

### Tested environment (Emulator? Real Device?)
Actual device. iPad mini 2

### Steps to reproduce the behavior

The code is as per pub example but with just slight modifications to include server receipt validation call. _buyProduct() is called on tap of the widget list item.

Future<Null> _buyProduct(IAPItem item) async {
    try {
      PurchasedItem purchased =
          await FlutterInappPurchase.buySubscription(item.productId);
      print('purchased: ${purchased.toString()}');     **<--- This never gets called**
      //send it to server for validation
      //https://github.com/Wizcorp/node-iap

      print('Updating user now....');
      if (this.mounted) {
        setState(() {
          _load = true;
        });
      }
      _verifyReceipt(purchased).then((verified) {
        if (verified == true) {
          _updateUser(purchased).then((futu) {
            if (this.mounted) {
              setState(() {
                _load = false;
                showInSnackBar('User updated with subscription details');
              });
            }
          });
        } else {
          //disable user subscription
        }
      });
    } catch (error) {
      print('_buyProduct() :: $error');
    }
  }

----------

Full log 

2018-09-05 06:02:03.678257+1000 Runner[1456:557945] [MC] System group container for systemgroup.com.apple.configurationprofiles path is /private/var/containers/Shared/SystemGroup/systemgroup.com.apple.configurationprofiles
2018-09-05 06:02:03.680136+1000 Runner[1456:557945] [MC] Reading from public effective user settings.
2018-09-05 06:02:03.709252+1000 Runner[1456:558047] flutter: result: true
2018-09-05 06:02:03.710976+1000 Runner[1456:558047] flutter: Platform is iOS 11.4.1
2018-09-05 06:02:03.712019+1000 Runner[1456:558047] flutter: consumeAllItems: no-ops in ios
2018-09-05 06:02:04.532263+1000 Runner[1456:558047] flutter: Item = productId: ####_monthly_###, price: 4.49, currency: AUD, localizedPrice: $4.49, title: #### Monthly Subscription, description: Monthly Subscription, introductoryPrice: , subscriptionPeriodNumberIOS: 1, subscriptionPeriodUnitIOS: MONTH, subscriptionPeriodAndroid: null, introductoryPriceCyclesAndroid: null, introductoryPricePeriodAndroid: null, freeTrialPeriodAndroid: null,
2018-09-05 06:02:09.887976+1000 Runner[1456:558047] flutter: -------------------------
2018-09-05 06:02:09.888190+1000 Runner[1456:558047] flutter: taped on subscription
2018-09-05 06:02:09.890083+1000 Runner[1456:557945] identifier ####_monthly_###
2018-09-05 06:02:09.891943+1000 Runner[1456:557945] 
 Purchase Started !!
Transaction in progress: <SKPaymentTransaction: 0x104#####>
2018-09-05 06:02:12.584199+1000 Runner[1456:557945] [App] if we're in the real pre-commit handler we can't actually add any new fences due to CA restriction
2018-09-05 06:02:13.706466+1000 Runner[1456:558066] TIC Read Status [6:0x0]: 1:57
2018-09-05 06:02:43.590435+1000 Runner[1456:557945] 

 Purchase Successful !! 
.
Transaction purchased or restored: <SKPaymentTransaction: 0x104#####>
warning: could not execute support code to read Objective-C class data in the process. This may reduce the quality of type information available.

**<Nothing is printed beyond this in the console>**

**Comment by vinvantest** [comment]
I am not sure PR? do you mean a pull request? I am not very good at this. I am pretty new to Git / flutter etc. Is it something you could do? I am a little uncomfortable doing that as I may screw things up at your end. It would be bad for the community who wish to use your great work.

**Comment by dooboolab** [comment]
@vinvantest Alright. I was just trying to give a chance to become a contributor to our repo. I'll update the code as in 24hours but if you learn how to give `PR`(pull request) before that, you may try anytime. It is very easy.

**Comment by vinvantest** [comment]
Once you update the code do I have to do flutter pub get to get the latest?

**Comment by vinvantest** [comment]
what version will it be?

**Comment by dooboolab** [comment]
@vinvantest I am looking for the fix now, but where should be the fix in the code? Below code isn't from our module.
```
func paymentQueue(_ queue: SKPaymentQueue, updatedTransactions transactions: [SKPaymentTransaction]) {
    var payments:Array<Dictionary<String, Any>> = []
    
    for transaction in transactions {
        switch transaction.transactionState {
        case .failed:
            queue.finishTransaction(transaction)
            print("Transaction Failed \(transaction)")
        case .purchased, .restored:
            queue.finishTransaction(transaction)
            print("Transaction purchased or restored: \(transaction)")
            payments.append(paymentToDictionary(transaction))
        case .deferred, .purchasing:
            print("Transaction in progress: \(transaction)")
        }
    }
    
    if(payments.count > 0) { . **<-- the fix**
        flutterResult?(payments)
    }
    else {
        flutterResult = nil
    }
}
```

**Comment by vinvantest** [comment]
You right.... I had flutter_payments package before attempting with your module and somehow it is in my .pub-cache. 

It doesn't clean the .pub-cache with $flutter clean command which I thought it should.

Do you also use the same file somewhere in your package?

SwiftFlutterPaymentsPlugin.swift
mac_installs/flutter/.pub-cache/hosted/pub.dartlang.org/flutter_payments-0.1.5/ios/Classes/SwiftFlutterPaymentsPlugin.swift

import Flutter
import UIKit
import StoreKit

public class SwiftFlutterPaymentsPlugin: NSObject, FlutterPlugin {
    var cachedProducts: [String: SKProduct] = [:]
    var productRequest: ProductRequestHandler?
    var transactionHandler: PaymentTransactionHandler
    
    override init() {
        transactionHandler = PaymentTransactionHandler()
        SKPaymentQueue.default().add(transactionHandler)
        
        super.init()
    }
    
     static let instance:SwiftFlutterPaymentsPlugin = SwiftFlutterPaymentsPlugin()
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "co.delightfulgoods.flutterpayments", binaryMessenger: registrar.messenger())
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let callArgs = call.arguments as? Dictionary<String, Any>

        switch call.method {
            case "getPurchaseHistory": getPurchaseHistory(result)
            case "getProducts": getProducts(result, callArgs!["skus"] as! [String])
            case "purchase": purchase(result, callArgs!["sku"] as! String)
            case "billingEnabled": result(billingEnabled)

            default: result(FlutterMethodNotImplemented)
        }
    }
    
    var billingEnabled: Bool {
        get { return SKPaymentQueue.canMakePayments(); }
    }
    
    public func purchase(_ result: FlutterResult, _ sku: String) {
        if let product = cachedProducts[sku] {
            let payment = SKPayment(product: product)
            SKPaymentQueue.default().add(payment)
        }
    }
    
    public func getPurchaseHistory(_ result: @escaping FlutterResult) {
        transactionHandler.flutterResult = result
        SKPaymentQueue.default().restoreCompletedTransactions()
    }
    
    public func getProducts(_ result: @escaping FlutterResult, _ skus: [String]) {
        let request = SKProductsRequest(productIdentifiers:  NSSet(array: skus) as! Set<String>)
        productRequest = ProductRequestHandler(result)
        request.delegate = productRequest
        request.start()
    }
}


class ProductRequestHandler : NSObject, SKProductsRequestDelegate {
    var flutterResult : FlutterResult?
    
    init(_ result: @escaping FlutterResult) {
        self.flutterResult = result;
    }
    
    func productsRequest(_ request: SKProductsRequest, didReceive response: SKProductsResponse) {
        flutterResult?(response.products.map(productToDictionary))
    }
    
    private func productToDictionary(_ productData:SKProduct) -> Dictionary<String, Any> {
        // SwiftFlutterPaymentsPlugin.instance.cachedProducts[productData.productIdentifier] = productData
        
        let numberFormatter = NumberFormatter()
        numberFormatter.formatterBehavior = .behavior10_4
        numberFormatter.numberStyle = .currency
        numberFormatter.locale = productData.priceLocale
    
        return [
            "sku": productData.productIdentifier,
            "title": productData.localizedTitle,
            "description": productData.localizedDescription,
            "price": numberFormatter.string(from: productData.price) ?? "",
        ]
    }
}


class PaymentTransactionHandler : NSObject, SKPaymentTransactionObserver {
    var flutterResult : FlutterResult?
    
    func paymentQueue(_ queue: SKPaymentQueue, updatedTransactions transactions: [SKPaymentTransaction]) {
        var payments:Array<Dictionary<String, Any>> = []
        
        for transaction in transactions {
            switch transaction.transactionState {
            case .failed:
                queue.finishTransaction(transaction)
                print("Transaction Failed \(transaction)")
            case .purchased, .restored:
                queue.finishTransaction(transaction)
                print("Transaction purchased or restored: \(transaction)")
                payments.append(paymentToDictionary(transaction))
            case .deferred, .purchasing:
                print("Transaction in progress: \(transaction)")
            }
        }
        
        if(payments.count > 0) {
            flutterResult?(payments)
        }
        else {
            flutterResult = nil
        }
    }
    
    private func paymentToDictionary(_ paymentData:SKPaymentTransaction) -> Dictionary<String, Any> {
        var purchaseTime:Int?
        
        if(paymentData.transactionDate != nil) {
            purchaseTime  = Int((paymentData.transactionDate!.timeIntervalSince1970 * 1000.0).rounded())
        }
        
        return [
            "orderId": paymentData.transactionIdentifier!,
            "packageName": paymentData.payment.productIdentifier,
            "purchaseTime": purchaseTime!,
            "sku": paymentData.payment.productIdentifier,
        ]
    }
}

**Comment by vinvantest** [comment]
Yes.. verified... the .pub-cache created issues. The issue is not related to your package.

**Comment by adriancmurray** [comment]
@vinvantest what was the work around for this? It's really only necessary for development but could also be an issue if/when a user switches itunes accounts on a phone. Not sure that happens often but it could potentially throw an issue.
